### PR TITLE
Add xml-reader subproject, a StAX based xml-reading library

### PR DESCRIPTION
**Pull request by DanVanAtta**
StAX is a DOM-less XML parser. StAX is essentially a cursor that
traverses an XML document and notifies the code of parsing events
as they happen.

This update adds a subproject, 'xml-reader' that wraps this library
and provides an annotation framework (very similar to JaxB) where
Java objects that model the XML can be annotated and automatically
created by 'XmlMapper'.

The sub-project README has details and documentation on how to set
this up.

As a feature the mapper can automatically map 'int', 'double'
and 'boolean' objects.

Of note, the StaX parser does not validate XML against a DTD.
If we want that validation we'll have to do so as an auxilary step
(perhaps something we can wire up for map-makers to validate
their XMLs).

Overall the sub-project is designed to give us a lot of control
over XML parsing. For example a logical next step will be to
add 'Aliases' to the annotations to support different naming of
equivalent tags and attributes.

The xml-reader is also designed to be efficient (fast) by using
StAX and emphasis has been put on yielding good error messages
when things go wrong.

As a larger fit, the XML-reader will be used to create data
model objects of the Game XMLs. These data model objects
will represent the data in XML almost exactly and then
will be used in another stage for 'game data assembly' where
they are put together to create a game data object.
This is different from 'GameParser' of today that parses
XML and directly creates a GameData object. This extra
layer should allow a lot of flexibility for different
XML formats, notably to support both future and legacy types.

## Additional Notes to Reviewer

This is a large and relatively polished sub-chunk of the draft PR: https://github.com/triplea-game/triplea/pull/7506

### CODE REVIEW

File: xml-reader/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+Usage:
+```java
+InputStream xmlInputStream = ... < open file input stream >
+XmlMapper mapper = new XmlMapper(xmlInputStream);
+Library library = mapper.mapXmlToObject(mapper);
+```
+
+## General Notes
+
+- XML tags are matched to model object class names (case insensitive)
```
Comment (DanVanAtta) [inline xml-reader/README.md:73]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Coverage increased per file
===========================
+ game-core/src/main/java/games/strategy/triplea/Properties.java  1
+ game-core/src/main/java/games/strategy/triplea/odds/calculator/DummyPlayer.java  8
         

Complexity increasing per file
==============================
- xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java  12
- xml-reader/src/test/java/org/triplea/generic/xml/reader/AttributeDataCastingTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/AbstractXmlMapperTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java  37
- xml-reader/src/test/java/org/triplea/generic/xml/reader/JavaDataModelErrorTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/XmlMapperTestUtils.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/BodyTextParsingTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/ExtendedLibraryExampleTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlParser.java  10
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/XmlParsingException.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/XmlDataException.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/XmlDataErrorTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/ReflectionUtils.java  9
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/JavaDataModelException.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java  8
- xml-reader/src/test/java/org/triplea/generic/xml/reader/SimpleLibraryExampleTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586&bid=20050961)

File: xml-reader/README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+- XML attributes are matched to model object attributes by variable name (case insensitive)
+- Model POJO objects must have a no-arg constructor.
```
Comment (DanVanAtta) [inline xml-reader/README.md:75]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+```java
+InputStream xmlInputStream = ... < open file input stream >
+XmlMapper mapper = new XmlMapper(xmlInputStream);
+Library library = mapper.mapXmlToObject(mapper);
+```
+
+## General Notes
+
+- XML tags are matched to model object class names (case insensitive)
+- XML attributes are matched to model object attributes by variable name (case insensitive)
```
Comment (DanVanAtta) [inline xml-reader/README.md:74]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        && field.getType() == Integer.class
+        && field.getAnnotation(Attribute.class).defaultInt() == 0) {
+      return null;
+    } else if (attributeValue == null) {
+      return field.getAnnotation(Attribute.class).defaultInt();
+    } else {
+      try {
+        return Integer.valueOf(attributeValue);
+      } catch (final NumberFormatException e) {
+        throw new XmlDataException(
```
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java:50]: Codacy found an issue: [New exception is thrown in catch block, original stack trace may be lost](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ <https://en.wikipedia.org/wiki/StAX>
+
+## Using the Xml-Reader, an Example
+
+Mapping XML to POJO objects is done by creating annotated model objects.
+The parser library will then use reflection to map XML elements encountered
+(tags, attributes, and body content) onto the annotated model objects.
+
+
+### Example
```
Comment (DanVanAtta) [inline xml-reader/README.md:16]: Codacy found an issue: [[no-consecutive-blank-lines] Remove 1 line before node](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                throw new JavaDataModelException(field, "Unexpected illegal access", e);
+              }
+            });
+      }
+
+      tagParser.parse(xmlStreamReader);
+
+      return instance;
+    } catch (final Throwable e) {
+      if (e instanceof XmlParsingException) {
```
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java:114]: Codacy found an issue: [An instanceof check is being performed on the caught exception.  Create a separate catch clause for this exception type.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+          "Too many body text fields, can only have one on any given class");
+    }
+    if (!bodyTextFields.isEmpty() && (!tagFields.isEmpty() && !tagListFields.isEmpty())) {
+      throw new JavaDataModelException(
+          "Illegal combination of annoations, may only have attributes and a body text,"
+              + "or attributes and tags (or taglist), but may not have both body text and tags.");
+    }
+  }
+
+  private static <T> void validateAnnotations(final Field field) throws JavaDataModelException {
```
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java:57]: Codacy found an issue: [The method 'validateAnnotations(Field)' has an NPath complexity of 10206, current threshold is 200](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+XmlMapper mapper = new XmlMapper(xmlInputStream);
+Library library = mapper.mapXmlToObject(mapper);
+```
+
+## General Notes
+
+- XML tags are matched to model object class names (case insensitive)
+- XML attributes are matched to model object attributes by variable name (case insensitive)
+- Model POJO objects must have a no-arg constructor.
+- Variables annotated with `@BodyText` can have any name
```
Comment (DanVanAtta) [inline xml-reader/README.md:76]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+- List parsing maintains XML ordering..
```
Comment (DanVanAtta) [inline xml-reader/README.md:77]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Coverage increased per file
===========================
+ game-core/src/main/java/games/strategy/triplea/Properties.java  1
+ game-core/src/main/java/games/strategy/triplea/odds/calculator/DummyPlayer.java  8
         

Complexity increasing per file
==============================
- xml-reader/src/test/java/org/triplea/generic/xml/reader/XmlMapperTestUtils.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/SimpleLibraryExampleTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/AttributeDataCastingTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java  12
- xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlParser.java  10
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/JavaDataModelException.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java  37
- xml-reader/src/test/java/org/triplea/generic/xml/reader/ExtendedLibraryExampleTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/XmlParsingException.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/JavaDataModelErrorTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/ReflectionUtils.java  9
- xml-reader/src/main/java/org/triplea/generic/xml/reader/exceptions/XmlDataException.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/AbstractXmlMapperTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/BodyTextParsingTest.java  1
- xml-reader/src/test/java/org/triplea/generic/xml/reader/XmlDataErrorTest.java  1
- xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586&bid=20050961)

File: xml-reader/README.md (hunk 9 above)
Comment (DanVanAtta) [inline xml-reader/README.md:77]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java (hunk 7 above)
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/AnnotatedFields.java:57]: Codacy found an issue: [The method 'validateAnnotations(Field)' has an NPath complexity of 10206, current threshold is 200](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java (hunk 6 above)
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/XmlMapper.java:114]: Codacy found an issue: [An instanceof check is being performed on the caught exception.  Create a separate catch clause for this exception type.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md (hunk 1 above)
Comment (DanVanAtta) [inline xml-reader/README.md:73]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md (hunk 5 above)
Comment (DanVanAtta) [inline xml-reader/README.md:16]: Codacy found an issue: [[no-consecutive-blank-lines] Remove 1 line before node](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java (hunk 4 above)
Comment (DanVanAtta) [inline xml-reader/src/main/java/org/triplea/generic/xml/reader/AttributeValueCasting.java:50]: Codacy found an issue: [New exception is thrown in catch block, original stack trace may be lost](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md (hunk 2 above)
Comment (DanVanAtta) [inline xml-reader/README.md:75]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md (hunk 8 above)
Comment (DanVanAtta) [inline xml-reader/README.md:76]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

File: xml-reader/README.md (hunk 3 above)
Comment (DanVanAtta) [inline xml-reader/README.md:74]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6130586)

**Comment by RoiEXLab** [comment]
👍

**Pull request merged**